pkg/lib/authn/authenticator/service/service.go

```diff
@@ -370,8 +370,7 @@ func (s *Service) WithSpec(ai *authenticator.Info, spec *authenticator.Spec) (bo
 	switch ai.Type {
 	case model.AuthenticatorTypePassword:
 		a := ai.Password
-		plainPassword := spec.Password.PlainPassword
-		newAuth, err := s.Password.WithPassword(a, plainPassword)
+		newAuth, err := s.Password.WithSpec(a, spec.Password)
```

Comment:
I suggest we introduce a new method for changing password.

```
func (s *Service) UpdatePassword(ai *authenticator.info, options *UpdatePasswordOptions) (bool, *authenticator.Info, error)

type UpdatePasswordOptions struct {
  PlainPassword string
  PasswordHash string
  # If true, then expireAt is set to ExpireAt.
  # To set NULL, SetExpireAt=true and ExpireAt=nil
  # To keep unchanged, SetExpireAt=false
  SetExpireAt bool
  ExpireAt *time.Time
}
```

And then visit every place where WithSpec is used, and change them to call UpdatePassword if necessary.

In the original WithSpec, panic if password authenticator is received.

Reply:
Also rename current `ResetPassword` to `ResetPasswordByEndUser`

Reply:
@louischan-oursky Renamed.